File: frontend/src/domain/liveDgt.ts
// Official road data from the DGT (Spain's traffic authority), as the backend reads it from the DGT's
// National Access Point (DATEX II): forest-fire incidents and road or carriageway closures. Official
// data, shown next to HackFire's own suggestions and never mixed with them. Pure types and functions.

import type { GeoJsonGeometry } from './spread'

/** One DATEX II situation record, as served by the backend. */
export interface DgtRecordResponse {
  id: string
  road: string | null
  cause: string | null
  cause_detail: string | null
  management: string | null
  forest_fire: boolean
  closure: boolean
  since: string | null
  from: { lat: number; lon: number; km: number | null; municipality: string | null; province: string | null }
  to: { lat: number; lon: number } | null
  comments: string[]
}

/** GET /api/live/dgt. */
export interface DgtOverviewResponse {
  source: string
  published_at: string | null
  fetched_at: string
  stale: boolean
  forest_fires: DgtRecordResponse[]
  closures: DgtRecordResponse[]
}

/** The `dgt` block of GET /api/live/operations/{fireId}: the records near that fire. */
export interface DgtNearResponse {
  available: boolean
  stale: boolean
  fetched_at?: string
  near_m?: number
  records: DgtRecordResponse[]
}

/** What kind of record, for its label: a forest fire, or how the road is managed. */
export type DgtKind = 'forestFire' | 'roadClosed' | 'carriagewayClosures' | 'laneClosures' | 'other'

/** Why a closure is in force, for its label; anything the dashboard does not name is `other`. */
export type DgtCause = 'roadMaintenance' | 'accident' | 'vehicleObstruction' | 'obstruction' | 'environmentalObstruction' | 'infrastructureDamageObstruction' | 'poorEnvironment' | 'publicEvent' | 'other'

const KNOWN_MANAGEMENT: readonly DgtKind[] = ['roadClosed', 'carriagewayClosures', 'laneClosures']
const KNOWN_CAUSES: readonly DgtCause[] = [
  'roadMaintenance',
  'accident',
  'vehicleObstruction',
  'obstruction',
  'environmentalObstruction',
  'infrastructureDamageObstruction',
  'poorEnvironment',
  'publicEvent',
]

export interface DgtRecord {
  id: string
  road: string | null
  kind: DgtKind
  /** The DGT shuts the road or a whole carriageway here, so the map draws it as an official cordon. */
  closure: boolean
  /** How the road is managed, when a forest fire's record also closes it. */
  management: DgtKind
  cause: DgtCause
  /** Since when, epoch ms; null when the DGT gives no start. */
  since: number | null
  lon: number
  lat: number
  km: number | null
  municipality: string | null
  province: string | null
  /** The far end of the stretch the record covers; null when the DGT gives only a point. */
  to: { lon: number; lat: number } | null
}

function management(value: string | null): DgtKind {
  return KNOWN_MANAGEMENT.find((known) => known === value) ?? 'other'
}

export function parseDgtRecord(record: DgtRecordResponse): DgtRecord {
  const since = record.since === null ? NaN : Date.parse(record.since)
  return {
    id: record.id,
    road: record.road,
    closure: record.closure,
    kind: record.forest_fire ? 'forestFire' : management(record.management),
    management: management(record.management),
    cause: KNOWN_CAUSES.find((known) => known === record.cause) ?? 'other',
    since: Number.isNaN(since) ? null : since,
    lon: record.from.lon,
    lat: record.from.lat,
    km: record.from.km,
    municipality: record.from.municipality,
    province: record.from.province,
    to: record.to === null ? null : { lon: record.to.lon, lat: record.to.lat },
  }
}

/** The records that shut a road or a carriageway: what the map draws as an official cordon. Pure. */
export function dgtClosures(records: readonly DgtRecord[]): DgtRecord[] {
  return records.filter((record) => record.closure)
}

/** Where a record is: the stretch the DGT gives, or its single point. Pure. */
export function dgtGeometry(record: DgtRecord): GeoJsonGeometry {
  if (record.to === null || (record.to.lon === record.lon && record.to.lat === record.lat)) {
    return { type: 'Point', coordinates: [record.lon, record.lat] }
  }
  return {
    type: 'LineString',
    coordinates: [
      [record.lon, record.lat],
      [record.to.lon, record.to.lat],
    ],
  }
}

export interface DgtOverview {
  forestFires: DgtRecord[]
  closures: DgtRecord[]
  fetchedAt: number
  stale: boolean
}

export function parseDgtOverview(response: DgtOverviewResponse): DgtOverview {
  return {
    forestFires: response.forest_fires.map(parseDgtRecord),
    closures: response.closures.map(parseDgtRecord),
    fetchedAt: Date.parse(response.fetched_at),
    stale: response.stale,
  }
}

export interface DgtNear {
  /** False when the DGT feed did not answer and nothing was cached. */
  available: boolean
  stale: boolean
  fetchedAt: number | null
  nearM: number | null
  records: DgtRecord[]
}

export function parseDgtNear(response: DgtNearResponse | undefined): DgtNear {
  if (!response) return { available: false, stale: false, fetchedAt: null, nearM: null, records: [] }
  const fetchedAt = response.fetched_at === undefined ? NaN : Date.parse(response.fetched_at)
  return {
    available: response.available,
    stale: response.stale,
    fetchedAt: Number.isNaN(fetchedAt) ? null : fetchedAt,
    nearM: response.near_m ?? null,
    records: response.records.map(parseDgtRecord),
  }
}

File: backend/app/live_dgt.py
"""Live mode's official road data: the DGT's forest-fire incidents and road closures across Spain.

Official data, published by the DGT (Spain's traffic authority) on its National Access Point as
DATEX II (providers/dgt.py). HackFire only reads it and shows it next to its own suggestions: the
forest-fire incidents on the map, and, for the selected fire, the DGT records within its simulated
footprint plus NEAR_FIRE_M.

The feed is about 3 MB, so the relevant records are cached in memory for CACHE_SECONDS and the last
good answer is written to CACHE_FILE. When the DGT does not answer, the last one comes back with
`stale: true`; with none, DgtUnavailable.
"""

import json
import threading
import time
from dataclasses import dataclass
from datetime import UTC, datetime

import httpx
from shapely.geometry import LineString, Point

from .config import DATA_DIR
from .providers import dgt

CACHE_FILE = DATA_DIR / "live_dgt.json"
CACHE_SECONDS = 300
# After a failure, the stale answer is served this long before the DGT is asked again.
RETRY_SECONDS = 60
TIMEOUT_SECONDS = 10
# DGT records this far from a fire's simulated footprint are listed with it.
NEAR_FIRE_M = 5000

SOURCE = "DGT"
SOURCE_NAME = "DGT, Punto de Acceso Nacional (DATEX II)"


class DgtUnavailable(Exception):
    """The DGT feed failed and nothing is cached."""


@dataclass
class _Snapshot:
    records: list[dict]
    published_at: str | None
    fetched_at: datetime
    expires: float  # time.monotonic()
    stale: bool = False


_snapshot: _Snapshot | None = None
_lock = threading.Lock()


def relevant(records: list[dict]) -> list[dict]:
    """Forest-fire incidents and road or carriageway closures, the only records live mode shows. Pure."""
    return [
        record
        for record in records
        if (record["forest_fire"] or record["closure"]) and record.get("validity") in (None, "active")
    ]


def _fetch() -> tuple[list[dict], str | None]:
    with httpx.Client(timeout=TIMEOUT_SECONDS) as client:
        feed = dgt.parse(dgt.fetch(client))
    return relevant(feed["records"]), feed["published_at"]


def _answer(snapshot: _Snapshot) -> dict:
    return {
        "source": SOURCE,
        "source_name": SOURCE_NAME,
        "published_at": snapshot.published_at,
        "fetched_at": snapshot.fetched_at.isoformat(),
        "stale": snapshot.stale,
        "records": snapshot.records,
    }


def _write_file(snapshot: _Snapshot) -> None:
    """The last good answer, for the next outage or restart. A read-only disk only loses the fallback."""
    try:
        CACHE_FILE.parent.mkdir(parents=True, exist_ok=True)
        temporary = CACHE_FILE.with_suffix(".tmp")
        temporary.write_text(json.dumps(_answer(snapshot), separators=(",", ":")), encoding="utf-8")
        temporary.replace(CACHE_FILE)
    except OSError:
        pass


def _read_file() -> _Snapshot | None:
    try:
        body = json.loads(CACHE_FILE.read_text(encoding="utf-8"))
        return _Snapshot(body["records"], body.get("published_at"), datetime.fromisoformat(body["fetched_at"]), 0.0)
    except (OSError, ValueError, KeyError, TypeError):
        return None


def snapshot() -> dict:
    """The DGT's forest-fire incidents and closures, at most CACHE_SECONDS old when the DGT answers.

    If the DGT fails, the last good answer (in memory, else from CACHE_FILE) comes back with
    `stale: true`. With neither, raises DgtUnavailable.
    """
    global _snapshot
    with _lock:
        now = time.monotonic()
        if _snapshot is None or now >= _snapshot.expires:
            try:
                records, published_at = _fetch()
                _snapshot = _Snapshot(records, published_at, datetime.now(UTC), now + CACHE_SECONDS)
            except (httpx.HTTPError, ValueError) as error:
                fallback = _snapshot or _read_file()
                if fallback is None:
                    raise DgtUnavailable(str(error)) from error
                _snapshot = _Snapshot(fallback.records, fallback.published_at, fallback.fetched_at, now + RETRY_SECONDS, stale=True)
            else:
                _write_file(_snapshot)
        return _answer(_snapshot)


def overview() -> dict:
    """GET /api/live/dgt: the whole of Spain, forest-fire incidents and closures apart."""
    body = snapshot()
    records = body.pop("records")
    return {
        **body,
        "forest_fires": [record for record in records if record["forest_fire"]],
        "closures": [record for record in records if record["closure"] and not record["forest_fire"]],
    }


def _shape(record: dict):
    start, end = record["from"], record.get("to")
    if end is None or (start["lon"], start["lat"]) == (end["lon"], end["lat"]):
        return Point(start["lon"], start["lat"])
    return LineString([(start["lon"], start["lat"]), (end["lon"], end["lat"])])


def near(records: list[dict], area) -> list[dict]:
    """The records whose point or stretch touches `area` (lon/lat), forest fires first, newest first. Pure."""
    found = [record for record in records if _shape(record).intersects(area)]
    found.sort(key=lambda record: record.get("since") or "", reverse=True)
    return sorted(found, key=lambda record: not record["forest_fire"])


def near_area(area) -> dict:
    """For one fire: the DGT records within `area`, or why there are none. Never raises."""
    try:
        body = snapshot()
    except DgtUnavailable:
        return {"source": SOURCE, "source_name": SOURCE_NAME, "available": False, "stale": False, "records": []}
    return {**body, "available": True, "near_m": NEAR_FIRE_M, "records": near(body["records"], area)}


def reset_cache() -> None:
    global _snapshot
    _snapshot = None
